### PR TITLE
feat: support RANGE in queries Part 2: Arrow

### PR DESCRIPTION
**Pull request by Linchin**
This PR supports getting RANGE values in queries as Arrow. This is the second part of the RANGE queries PR, and should not be merged until part 1 is merged. Part 1 #1884 supports RANGE as JSON.

Part of #1724🦕

### CODE REVIEW

File: tests/unit/test_table.py [hunk 1]
```diff
@@ -3503,7 +3503,10 @@ def test_to_dataframe_no_tqdm_no_progress_bar(self):
         user_warnings = [
             warning for warning in warned if warning.category is UserWarning
         ]
-        self.assertEqual(len(user_warnings), 0)
+        # Note: number of warnings is inconsistent across python versions
```
Comment (Linchin) [inline tests/unit/test_table.py:3506]: I removed the assertion for number of warnings, because we get different number of them with different python versions. Same for line 3540

File: tests/unit/test_table.py [hunk 2]
```diff
@@ -3534,7 +3537,10 @@ def test_to_dataframe_no_tqdm(self):
         user_warnings = [
             warning for warning in warned if warning.category is UserWarning
         ]
-        self.assertEqual(len(user_warnings), 1)
+        # Note: number of warnings is inconsistent across python versions
```
Comment (Linchin) [inline tests/unit/test_table.py:3540]: I removed the assertion for number of warnings, because we get different number of them with different python versions.

File: google/cloud/bigquery/_helpers.py [hunk 3]
```diff
@@ -309,6 +309,46 @@ def _json_from_json(value, field):
         return None
 
 
+def _range_element_from_json(value, field):
+    """Coerce 'value' to a range element value, if set or not nullable."""
+    if value == "UNBOUNDED":
+        return None
+    elif field.element_type == "DATE":
```
Comment (shollyman) [inline google/cloud/bigquery/_helpers.py:316]: Do we have a list of the supported range element types somewhere?  If so, this could like be simplified to something like
```
if field.element_type IN _SUPPORTED_RANGE_ELEMENTS:
  return _CELLDATA[field.element_type](value, None)
else:
  raise ValueError...
```

File: google/cloud/bigquery/_pyarrow_helpers.py [hunk 4]
```diff
@@ -68,6 +90,9 @@ def pyarrow_timestamp():
         "STRING": pyarrow.string,
         "TIME": pyarrow_time,
         "TIMESTAMP": pyarrow_timestamp,
+        "RANGE<TIMESTAMP>": pyarrow_range_timestamp,
```
Comment (shollyman) [inline google/cloud/bigquery/_pyarrow_helpers.py:93]: addmittedly I'm not super familiar with this code, but it seems like we're ending up declaring N scalar versions of range.  Do we get a simpler story from treating range more like a complex type and branching our struct processing logic?

File: google/cloud/bigquery/_helpers.py (hunk 3 above)
Comment (Linchin) [reply google/cloud/bigquery/_helpers.py]: This way looks so much neater! I updated the code.

File: google/cloud/bigquery/_pyarrow_helpers.py (hunk 4 above)
Comment (Linchin) [reply google/cloud/bigquery/_pyarrow_helpers.py:93]: Indeed. I changed the code and opted to add a new conversion function, parallel to scalar and struct. I chose this instead of using the existing function for struct, because RANGE has fixed field names, so the logic is simpler.

File: google/cloud/bigquery/_pandas_helpers.py [hunk 5]
```diff
@@ -142,6 +142,12 @@ def bq_to_arrow_struct_data_type(field):
     return pyarrow.struct(arrow_fields)
 
 
+def bq_to_arrow_range_data_type(field):
+    element_type = field.element_type.upper()
+    arrow_element_type = _pyarrow_helpers.bq_to_arrow_scalars(element_type)()
```
Comment (shollyman) [inline google/cloud/bigquery/_pandas_helpers.py:147]: do we need to do validation here?  None-check?

File: google/cloud/bigquery/_pandas_helpers.py [hunk 6]
```diff
@@ -274,6 +286,22 @@ def types_mapper(arrow_data_type):
         elif time_dtype is not None and pyarrow.types.is_time(arrow_data_type):
             return time_dtype
 
+        elif pyarrow.types.is_struct(arrow_data_type):
```
Comment (shollyman) [inline google/cloud/bigquery/_pandas_helpers.py:289]: Do we need to handle structs more generally here, or is that logic elsewhere?

File: google/cloud/bigquery/table.py [hunk 7]
```diff
@@ -2215,6 +2279,69 @@ def to_dataframe(
...
+                    pyarrow.struct(
+                        [("start", pyarrow.date32()), ("end", pyarrow.date32())]
+                    )
+                )
+            except AttributeError:
+                # pandas.ArrowDtype was introduced in pandas 1.5, but python 3.7
+                # only supports upto pandas 1.3. If pandas.ArrowDtype is not
+                # present, we raise a warning and set range_date_dtype to None.
+                msg = (
+                    "Unable ro find class ArrowDtype in pandas, setting "
```
Comment (shollyman) [inline google/cloud/bigquery/table.py:2294]: s/ro/to/ here and in the other two msgs below

File: google/cloud/bigquery_v2/types/model.py [hunk 8]
```diff
@@ -103,6 +103,7 @@ class Model(proto.Message):
 
     class ModelType(proto.Enum):
         r"""Indicates the type of the Model."""
+
```
Comment (shollyman) [inline google/cloud/bigquery_v2/types/model.py:106]: these formatting changes are benign, but it's not clear why they're in this PR.  Maybe pull out unrelated cleanups into a separate PR?

File: tests/unit/test_table.py [hunk 9]
```diff
@@ -3497,13 +3497,9 @@ def test_to_dataframe_no_tqdm_no_progress_bar(self):
         api_request = mock.Mock(return_value={"rows": rows})
         row_iterator = self._make_one(_mock_client(), api_request, path, schema)
 
-        with warnings.catch_warnings(record=True) as warned:
+        with warnings.catch_warnings(record=True):
             df = row_iterator.to_dataframe(create_bqstorage_client=False)
 
-        user_warnings = [
```
Comment (shollyman) [inline tests/unit/test_table.py:3503]: Not familiar with these tests, but are we losing some existing signal here?  Or did test expectations change somehow?  Its not clear how this related to the PR intent.

**Review by shollyman — COMMENTED** [review]
Thanks for putting this together. Mostly minor comments and questions.

File: google/cloud/bigquery_v2/types/model.py (hunk 8 above)
Comment (Linchin) [reply google/cloud/bigquery_v2/types/model.py:106]: Indeed, I will revert these changes

File: google/cloud/bigquery/table.py (hunk 7 above)
Comment (Linchin) [reply google/cloud/bigquery/table.py:2294]: Thanks for catching this!

File: google/cloud/bigquery/_pandas_helpers.py (hunk 5 above)
Comment (Linchin) [reply google/cloud/bigquery/_pandas_helpers.py:147]: Great point, I will add a None-check here

File: tests/unit/test_table.py (hunk 9 above)
Comment (Linchin) [reply tests/unit/test_table.py:3503]: Indeed, we are losing some signals here - we no longer validate the number of warnings here. With this PR we will have different length of warnings [depending on the version of pandas](https://github.com/Linchin/python-bigquery/blob/get-query-results-range/google/cloud/bigquery/table.py#L2301-L2343). I deleted the warnings check, because I didn't want to have pandas version hard-coded in our tests (Should I be concerned about this?). Another alternative I can think of is to check `self.assertTrue(len(user_warnings) in {length_with_older_pandas, length_with_newer_pandas})`- this way we will lose some signal but less. What do you think?

File: google/cloud/bigquery/_pandas_helpers.py (hunk 5 above)
Comment (Linchin) [reply google/cloud/bigquery/_pandas_helpers.py:152]: I added it, as well as the unit tests in `test__pandas_helpers.py`.

File: google/cloud/bigquery/_pandas_helpers.py (hunk 6 above)
Comment (Linchin) [reply google/cloud/bigquery/_pandas_helpers.py:294]: Good question! Indeed, our types mapper function doesn't seem to do any conversion for STRUCT or ARRAY. This function is used as the parameter `types_mapper` to [pyarrow's `Table.to_pandas()`](https://arrow.apache.org/docs/python/generated/pyarrow.Table.html#pyarrow.Table.to_pandas), allowing for customizable type mapping from pyarrow to pandas. I'm not entirely sure why we didn't provide struct/array mapping options, but I think it might be related to the fact that the fields of a struct can be of any type, so the conversion can become quite complicated.

File: google/cloud/bigquery/_helpers.py [hunk 10]
```diff
@@ -310,17 +312,13 @@ def _json_from_json(value, field):
...
-        return _date_from_json(value, None)
-    elif field.element_type == "DATETIME":
-        return _datetime_from_json(value, None)
-    elif field.element_type == "TIMESTAMP":
-        return _timestamp_from_json(value, None)
+    if field.element_type in _SUPPORTED_RANGE_ELEMENTS:
+        return _CELLDATA_FROM_JSON[field.element_type](value, field.element_type)
     else:
-        raise ValueError(f"Unsupported range field type: {value}")
+        raise ValueError(f"Unsupported range field type: {field.element_type}")
```
Comment (shollyman) [inline google/cloud/bigquery/_helpers.py:321]: nit: should we change this to indicate the element type is unsupported, rather than "field type"?

File: tests/unit/test_table.py (hunk 9 above)
Comment (shollyman) [reply tests/unit/test_table.py:3503]: If all the supported versions emit at least one UserWarning we could still do an assert on the count being nonzero, but if we don't think this is a useful validation I'm fine removing it as well.

**Review by shollyman — APPROVED** [review]
Left some followup comments and a very minor nit, otherwise LGTM.  Thanks!

File: google/cloud/bigquery/_helpers.py (hunk 10 above)
Comment (Linchin) [reply google/cloud/bigquery/_helpers.py:321]: Indeed, I'll change it to be consistent with the name of the field.

**Comment by Linchin** [comment]
> If all the supported versions emit at least one UserWarning we could still do an assert on the count being nonzero, but if we don't think this is a useful validation I'm fine removing it as well.

Indeed, I added back the test, and it will pass when `len(warnings) == 0` or `len(warnings) == 3`, so we can cover both cases. I added a comment that explains why and when each value should be true. (Same for another warnings number check).
